pcsx2/IPC.cpp

```diff
@@ -0,0 +1,371 @@
+			// use a bunch of auto
+			auto receive_length = 0;
+			auto end_length = 4;
+
+			// while we haven't received the entire packet, maybe due to
+			// socket datagram splittage, we continue to read
+			while (receive_length < end_length)
+			{
+				auto tmp_length = read_portable(m_msgsock, &m_ipc_buffer[receive_length], MAX_IPC_SIZE - receive_length);
+
+				// we close the connection if an error happens
+				if (tmp_length <= 0)
+				{
+					receive_length = 0;
+					break;
+				}
+
+				receive_length += tmp_length;
+
+				// if we got at least the final size then update
+				if (end_length == 4 && receive_length >= 4)
+				{
+					end_length = FromArray<u32>(m_ipc_buffer, 0);
+					// we'd like to avoid a client trying to do OOB
+					if (end_length > MAX_IPC_SIZE)
+					{
+						receive_length = 0;
+						break;
+					}
+				}
+			}
+			SocketIPC::IPCBuffer res;
+
+			// we remove 4 bytes to get the message size out of the IPC command
+			// size in ParseCommand
+			if (receive_length == 0)
+				res = IPCBuffer{5, MakeFailIPC(m_ret_buffer)};
+			else
+				res = ParseCommand(&m_ipc_buffer[4], m_ret_buffer, (u32)receive_length - 4);
+
+			// we don't care about the error value as we will reset the
+			// connection after that anyways
+			if (write_portable(m_msgsock, res.buffer, res.size) < 0)
+			{
+			}
+		}
+		close_portable(m_msgsock);
+	}
+	return;
+}
+
+SocketIPC::~SocketIPC()
+{
+	m_end = true;
+#ifdef _WIN32
+	WSACleanup();
+#else
+	unlink(m_socket_name);
+#endif
+	close_portable(m_sock);
+	close_portable(m_msgsock);
+	delete[] m_ret_buffer;
+	delete[] m_ipc_buffer;
+	// destroy the thread
+	try
+	{
+		pxThread::Cancel();
+	}
+	DESTRUCTOR_CATCHALL
+}
+
+SocketIPC::IPCBuffer SocketIPC::ParseCommand(char* buf, char* ret_buffer, u32 buf_size)
+{
+	// currently all our instructions require a running VM so we check once
+	// here, slightly helps performance
+	if (!m_vm->HasActiveMachine())
+		return IPCBuffer{5, MakeFailIPC(ret_buffer)};
+
+	u32 ret_cnt = 5;
+	u32 buf_cnt = 0;
+
+	while (buf_cnt < buf_size)
+	{
+		// if we haven't received enough byte for the address used in R/W
+		// commands and opcode, changeme when address is out of the header!
+		if (!SafetyChecks(buf_cnt, 4 + 1, ret_cnt, 0, buf_size))
+			return IPCBuffer{5, MakeFailIPC(ret_buffer)};
+
+		// YY YY YY YY from schema below
+		// curently always used by implemented commands so it is out of the
+		// loop
+		const u32 a = FromArray<u32>(&buf[buf_cnt], 1);
+
+		//         IPC Message event (1 byte)
+		//         |  Memory address (4 byte)
+		//         |  |           argument (VLE)
+		//         |  |           |
+		// format: XX YY YY YY YY ZZ ZZ ZZ ZZ
+		//        reply code: 00 = OK, FF = NOT OK
+		//        |  return value (VLE)
+		//        |  |
+		// reply: XX ZZ ZZ ZZ ZZ
+		switch ((IPCCommand)buf[buf_cnt])
+		{
+			case MsgRead8:
+			{
+				if (!SafetyChecks(buf_cnt, 5, ret_cnt, 1, buf_size))
+					goto error;
+				u8 res;
```

Comment:
Maybe `const u8 res = memRead8(a);` instead?
Same for the others.

Reply:
Done!